File: 987/987.kt
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/987.py

import java.math.BigInteger

val WINDOWS: Array<IntArray> = arrayOf(
    intArrayOf(0, 1, 2, 3, 4),
    intArrayOf(1, 2, 3, 4, 5),
    intArrayOf(2, 3, 4, 5, 6),
    intArrayOf(3, 4, 5, 6, 7),
    intArrayOf(4, 5, 6, 7, 8),
    intArrayOf(5, 6, 7, 8, 9),
    intArrayOf(6, 7, 8, 9, 10),
    intArrayOf(7, 8, 9, 10, 11),
    intArrayOf(8, 9, 10, 11, 12),
    intArrayOf(9, 10, 11, 12, 0),
)

val OVERLAP: Array<BooleanArray> = Array(10) { BooleanArray(10) }

val PERMS: Array<Array<BigInteger>> = Array(5) { Array(5) { BigInteger.ZERO } }

fun initGlobals() {
    for (i in 0 until 10) {
        val setI = WINDOWS[i].toHashSet()
        for (j in 0 until 10) {
            OVERLAP[i][j] = WINDOWS[j].any { it in setI }
        }
    }
    for (n in 0 until 5) {
        PERMS[n][0] = BigInteger.ONE
        var value = BigInteger.ONE
        for (k in 1 until 5) {
            if (k <= n) {
                value = value * BigInteger.valueOf((n - (k - 1)).toLong())
                PERMS[n][k] = value
            }
        }
    }
}

fun coloringsOfAllSubsets(starts: IntArray): Array<BigInteger> {
    val k = starts.size
    val full = 1 shl k
    val adjacency = IntArray(k)
    for (i in 0 until k) {
        for (j in i + 1 until k) {
            if (OVERLAP[starts[i]][starts[j]]) {
                adjacency[i] = adjacency[i] or (1 shl j)
                adjacency[j] = adjacency[j] or (1 shl i)
            }
        }
    }
    val independent = BooleanArray(full)
    independent[0] = true
    for (mask in 1 until full) {
        val bit = mask and (-mask)
        val vertex = Integer.numberOfTrailingZeros(bit)
        val rest = mask xor bit
        independent[mask] = independent[rest] && ((adjacency[vertex] and rest) == 0)
    }
    var dp = Array(full) { BigInteger.ZERO }
    dp[0] = BigInteger.ONE
    repeat(4) {
        val newDp = Array(full) { BigInteger.ZERO }
        for (mask in 0 until full) {
            var total = BigInteger.ZERO
            var sub = mask
            while (true) {
                if (independent[sub]) {
                    total = total + dp[mask xor sub]
                }
                if (sub == 0) break
                sub = (sub - 1) and mask
            }
            newDp[mask] = total
        }
        dp = newDp
    }
    return dp
}

fun popcount(x: Int): Int = Integer.bitCount(x)

fun labeledCount(starts: IntArray): BigInteger {
    val k = starts.size
    val full = 1 shl k
    val colorings = coloringsOfAllSubsets(starts)
    val totalActive = IntArray(13)
    val activeMasksByRank = IntArray(13)
    for (index in starts.indices) {
        val bit = 1 shl index
        for (rank in WINDOWS[starts[index]]) {
            totalActive[rank] += 1
            activeMasksByRank[rank] = activeMasksByRank[rank] or bit
        }
    }
    var total = BigInteger.ZERO
    for (mask in 0 until full) {
        var ways = BigInteger.ONE
        for (rank in 0 until 13) {
            val monochromaticHere = popcount(activeMasksByRank[rank] and mask)
            val flexibleHere = totalActive[rank] - monochromaticHere
            val waysAtRank = PERMS[4 - monochromaticHere][flexibleHere]
            if (waysAtRank == BigInteger.ZERO) {
                ways = BigInteger.ZERO
                break
            }
            ways = ways * waysAtRank
        }
        val term = colorings[mask] * ways
        if (popcount(mask) and 1 == 1) {
            total = total - term
        } else {
            total = total + term
        }
    }
    return total
}

fun factorial(n: Int): BigInteger {
    var r = BigInteger.ONE
    for (i in 2..n) r = r * BigInteger.valueOf(i.toLong())
    return r
}

fun feasibleTypeCounts(target: Int): List<IntArray> {
    val results = mutableListOf<IntArray>()
    val counts = IntArray(10)
    val coverage = IntArray(13)
    fun backtrack(pos: Int, remaining: Int) {
        if (pos == 10) {
            if (remaining == 0) results.add(counts.copyOf())
            return
        }
        for (amount in 0..remaining) {
            var ok = true
            for (rank in WINDOWS[pos]) {
                coverage[rank] += amount
                if (coverage[rank] > 4) ok = false
            }
            counts[pos] = amount
            if (ok) backtrack(pos + 1, remaining - amount)
            for (rank in WINDOWS[pos]) {
                coverage[rank] -= amount
            }
        }
        counts[pos] = 0
    }
    backtrack(0, target)
    return results
}

fun countDisjointStraights(target: Int): BigInteger {
    var total = BigInteger.ZERO
    for (typeCounts in feasibleTypeCounts(target)) {
        val starts = mutableListOf<Int>()
        var divisor = BigInteger.ONE
        for (start in typeCounts.indices) {
            val amount = typeCounts[start]
            repeat(amount) { starts.add(start) }
            divisor = divisor * factorial(amount)
        }
        total = total + labeledCount(starts.toIntArray()) / divisor
    }
    return total
}

fun main() {
    initGlobals()
    check(countDisjointStraights(1) == BigInteger.valueOf(10200))
    check(countDisjointStraights(2) == BigInteger.valueOf(31832952))
    println(countDisjointStraights(8))
}

File: 987/987.rs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/987.py

const WINDOWS: [[usize; 5]; 10] = [
    [0, 1, 2, 3, 4],
    [1, 2, 3, 4, 5],
    [2, 3, 4, 5, 6],
    [3, 4, 5, 6, 7],
    [4, 5, 6, 7, 8],
    [5, 6, 7, 8, 9],
    [6, 7, 8, 9, 10],
    [7, 8, 9, 10, 11],
    [8, 9, 10, 11, 12],
    [9, 10, 11, 12, 0],
];

fn build_overlap() -> [[bool; 10]; 10] {
    let mut overlap = [[false; 10]; 10];
    for i in 0..10 {
        for j in 0..10 {
            let mut found = false;
            'outer: for &r in &WINDOWS[i] {
                for &s in &WINDOWS[j] {
                    if r == s {
                        found = true;
                        break 'outer;
                    }
                }
            }
            overlap[i][j] = found;
        }
    }
    overlap
}

fn build_perms() -> [[i64; 5]; 5] {
    let mut perms = [[0i64; 5]; 5];
    for n in 0..5 {
        perms[n][0] = 1;
        let mut value = 1i64;
        for k in 1..5 {
            if k <= n {
                value *= (n - (k - 1)) as i64;
                perms[n][k] = value;
            }
        }
    }
    perms
}

fn colorings_of_all_subsets(starts: &[usize], overlap: &[[bool; 10]; 10]) -> Vec<i64> {
    let k = starts.len();
    let full = 1usize << k;
    let mut adjacency = vec![0usize; k];
    for i in 0..k {
        for j in (i + 1)..k {
            if overlap[starts[i]][starts[j]] {
                adjacency[i] |= 1 << j;
                adjacency[j] |= 1 << i;
            }
        }
    }
    let mut independent = vec![false; full];
    independent[0] = true;
    for mask in 1..full {
        let bit = mask & mask.wrapping_neg();
        let vertex = bit.trailing_zeros() as usize;
        let rest = mask ^ bit;
        independent[mask] = independent[rest] && ((adjacency[vertex] & rest) == 0);
    }
    let mut dp = vec![0i64; full];
    dp[0] = 1;
    for _ in 0..4 {
        let mut new_dp = vec![0i64; full];
        for mask in 0..full {
            let mut total = 0i64;
            let mut sub = mask;
            loop {
                if independent[sub] {
                    total += dp[mask ^ sub];
                }
                if sub == 0 {
                    break;
                }
                sub = (sub - 1) & mask;
            }
            new_dp[mask] = total;
        }
        dp = new_dp;
    }
    dp
}

fn labeled_count(starts: &[usize], overlap: &[[bool; 10]; 10], perms: &[[i64; 5]; 5]) -> i64 {
    let k = starts.len();
    let full = 1usize << k;
    let colorings = colorings_of_all_subsets(starts, overlap);
    let mut total_active = [0usize; 13];
    let mut active_masks_by_rank = [0usize; 13];
    for (index, &start) in starts.iter().enumerate() {
        let bit = 1 << index;
        for &rank in &WINDOWS[start] {
            total_active[rank] += 1;
            active_masks_by_rank[rank] |= bit;
        }
    }
    let mut total = 0i64;
    for mask in 0..full {
        let mut ways = 1i64;
        for rank in 0..13 {
            let monochromatic_here = (active_masks_by_rank[rank] & mask).count_ones() as usize;
            let flexible_here = total_active[rank] - monochromatic_here;
            let ways_at_rank = perms[4 - monochromatic_here][flexible_here];
            if ways_at_rank == 0 {
                ways = 0;
                break;
            }
            ways *= ways_at_rank;
        }
        let term = colorings[mask] * ways;
        if mask.count_ones() & 1 == 1 {
            total -= term;
        } else {
            total += term;
        }
    }
    total
}

fn factorial(n: i64) -> i64 {
    if n <= 1 { 1 } else { n * factorial(n - 1) }
}

fn backtrack(
    pos: usize,
    remaining: usize,
    counts: &mut [usize; 10],
    coverage: &mut [usize; 13],
    results: &mut Vec<[usize; 10]>,
) {
    if pos == 10 {
        if remaining == 0 {
            results.push(*counts);
        }
        return;
    }
    for amount in 0..=(remaining) {
        let mut ok = true;
        for &rank in &WINDOWS[pos] {
            coverage[rank] += amount;
            if coverage[rank] > 4 {
                ok = false;
            }
        }
        counts[pos] = amount;
        if ok {
            backtrack(pos + 1, remaining - amount, counts, coverage, results);
        }
        for &rank in &WINDOWS[pos] {
            coverage[rank] -= amount;
        }
    }
    counts[pos] = 0;
}

fn feasible_type_counts(target: usize) -> Vec<[usize; 10]> {
    let mut results = Vec::new();
    let mut counts = [0usize; 10];
    let mut coverage = [0usize; 13];
    backtrack(0, target, &mut counts, &mut coverage, &mut results);
    results
}

fn count_disjoint_straights(target: usize) -> i64 {
    let overlap = build_overlap();
    let perms = build_perms();
    let mut total = 0i64;
    for type_counts in feasible_type_counts(target) {
        let mut starts = Vec::new();
        let mut divisor = 1i64;
        for (start, &amount) in type_counts.iter().enumerate() {
            for _ in 0..amount {
                starts.push(start);
            }
            divisor *= factorial(amount as i64);
        }
        total += labeled_count(&starts, &overlap, &perms) / divisor;
    }
    total
}

fn main() {
    assert_eq!(count_disjoint_straights(1), 10200);
    assert_eq!(count_disjoint_straights(2), 31832952);
    println!("{}", count_disjoint_straights(8));
}
